analyses/bangladesh/trigger_development/bgd-trigger-part2.md

```diff
@@ -0,0 +1,363 @@
+    TP = np.empty(nthresh)
+    FN = np.empty(nthresh)
+    FP = np.zeros(nthresh)
+    # Drop any NAs in the glofas columns
+    df = df_final[['observed', 'event', glofas_var_name]]
+    df = df[df[glofas_var_name].notna()]
+    for ithresh, thresh in enumerate(thresh_array):
+        detections = get_glofas_detections(df[glofas_var_name], thresh)
+        detection_window_min = 0
+        detection_window_max = 7
+        detected_event_dates = []
+        for detection in detections:
+            detected_events = df['event'][detection:
+                                    detection+GLOFAS_DETECTION_WINDOW]
+            detected_event_dates += (
+                list(detected_events[detected_events==True].index))
+            if sum(detected_events) == 0:
+                FP[ithresh] += 1
+        # Get the unique event dates
+        detected_event_dates = list(set(detected_event_dates))
+        event_dates = df.index[df['event']]
+        events_are_detected = [event_date in detected_event_dates for event_date in event_dates]
+        TP[ithresh] = sum(events_are_detected)
+        FN[ithresh] = sum([not event_detected for event_detected in events_are_detected])
+    return TP, FP, FN
+
+def get_more_stats(TP, FP, FN):
+    precision = TP / (TP + FP)
+    recall = TP / (TP + FN)
+    f1 = 2 / ((1/recall) + (1/precision))
+    return precision, recall, f1
+
+
+```
+
+```python
+def plot_stats(glofas_var_name, thresh_array, x_axis_units='[m$^3$ s$^{-1}$]'):
+    TP, FP, FN = get_detection_stats(glofas_var_name,
+                                thresh_array)
+    precision, recall, f1 = get_more_stats(TP, FP, FN)
+
+    # Plot results
+    x = thresh_array
+    fig, ax = plt.subplots()
+    ax.plot(x, TP, label='TP')
+    ax.plot(x, FP, label='FP')
+    #ax.plot(x, FN, label='FN')
+    df = df_final[['observed', 'event', glofas_var_name]]
+    df = df[df[glofas_var_name].notna()]
+    ax.axhline(len(df[df['event']]), c='C0', ls='--')
+    #ax.axhline(0, c='C2', ls='--')
+    #ax2 = ax.twinx()
+    #ax2.plot(x, precision, label='precision', c='r')
+    #ax2.plot(x, recall, label='recall', c='y')
+    #ax2.plot(x, f1, label='F1', c='k')
+    #ax2.set_ylim(-0.05, 1.05)
+    ax.legend(loc=2)
+    #ax2.legend(loc=0)
+    #ax2.axhline(0.2, c='r', ls='--')
+
+    ax.set_xlabel(f'GloFAS trigger threshold {x_axis_units}')
+    ax.set_ylabel('Number')
+    
+    
+def get_thresh_and_max_precision_with_no_fn( glofas_var_name, thresh_array):
+    TP, FP, FN = get_detection_stats(glofas_var_name,
+                                thresh_array)
+    precision, recall, f1 = get_more_stats(TP, FP, FN)
+    print('threshold', thresh_array[FN==0][-5:])
+    print('precision', precision[FN==0][-5:])
+    print('FP', FP[FN==0][-5:])
+```
+
+```python
+thresh_array = np.arange(19.0, 22.0, 0.05)
+glofas_var_name = 'wl_estimate'
+plot_stats(glofas_var_name, thresh_array, x_axis_units='[m]')
+get_thresh_and_max_precision_with_no_fn(glofas_var_name, thresh_array)
```

Comment:
Yes, because this method is specifically looking for threshold, precision, and FP around where TP reaches the maximum value. However, that does not occur in this case.